File: Problemset/0814-Binary Tree Pruning/Recursion.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {TreeNode}
def prune_tree(root)
    return nil if root.nil?

    root.left = prune_tree(root.left)
    root.right = prune_tree(root.right)

    if root.val == 1 or not root.left.nil? or not root.right.nil?
        return root
    else
        return nil
    end
end

File: Problemset/0144-Binary Tree Preorder Traversal/Recursion.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        return [root.val] + \
            self.preorderTraversal(root.left) + \
            self.preorderTraversal(root.right)
